### ISSUE TITLE
[tf.keras] Stateful Metrics assorted errors.

### ISSUE THREAD

**Issue by brge17** [tensorflow/tensorflow#20529]
I will break this issue down into several code snippets each displaying a different error. @fchollet. In total 3 issues. All of these issues are only relevant to ``tf.keras`` implementation. The ``keras`` implementation works as intended.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary.
- **TensorFlow version (use command below)**: 1.9.0
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: n/a

### Problem 1
Issues with multi-input/multi-output and batch averaging. This happens for both train and validation metrics.

### Source code/logs

```python
import tensorflow as tf

from tensorflow.python.keras.datasets import mnist
from tensorflow.python.keras.models import Model
from tensorflow.python.keras.layers import Input, Conv2D, MaxPooling2D, Flatten, Dense, UpSampling2D


class BatchCounter(tf.keras.layers.Layer):

        def __init__(self, name='batch_counter', **kwargs):
            super(BatchCounter, self).__init__(name=name, **kwargs)
            self.stateful = True
            self.batches = tf.keras.backend.variable(value=0, dtype='int32')

        def reset_states(self):
            tf.keras.backend.set_value(self.batches, 0)

        def __call__(self, y_true, y_pred):
            updates = [tf.keras.backend.update_add(self.batches, tf.keras.backend.variable(value=1, dtype='int32'))]
            self.add_update(updates)
            return self.batches


batch_size = 100
num_classes = 10
epochs = 1

# input image dimensions
img_rows, img_cols = 28, 28

# Data
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1).astype('float32') / 255
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1).astype('float32') / 255
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

# Convolutional Encoder
input_img = Input(shape=(img_rows, img_cols, 1))
conv_1 = Conv2D(16, (3, 3), activation='relu', padding='same')(input_img)
pool_1 = MaxPooling2D((2, 2), padding='same')(conv_1)
conv_2 = Conv2D(8, (3, 3), activation='relu', padding='same')(pool_1)
pool_2 = MaxPooling2D((2, 2), padding='same')(conv_2)
conv_3 = Conv2D(8, (3, 3), activation='relu', padding='same')(pool_2)
encoded= MaxPooling2D((2, 2), padding='same')(conv_3)

# Classification
flatten = Flatten()(encoded)
fc = Dense(128, activation='relu')(flatten)
softmax = Dense(num_classes, activation='softmax', name='classification')(fc)

# Decoder
conv_4 = Conv2D(8, (3, 3), activation='relu', padding='same')(encoded)
up_1 = UpSampling2D((2, 2))(conv_4)
conv_5 = Conv2D(8, (3, 3), activation='relu', padding='same')(up_1)
up_2 = UpSampling2D((2, 2))(conv_5)
conv_6 = Conv2D(16, (3, 3), activation='relu')(up_2)
up_3 = UpSampling2D((2, 2))(conv_6)
decoded = Conv2D(1, (3, 3), activation='sigmoid', padding='same', name='autoencoder')(up_3)

model = Model(inputs=input_img, outputs=[softmax, decoded])

model.compile(loss={'classification': 'categorical_crossentropy',
                    'autoencoder': 'binary_crossentropy'},
              loss_weights={'classification': 1.0,
                            'autoencoder': 0.5},
              optimizer='adam',
              metrics={'classification': 'accuracy', 'autoencoder': BatchCounter()})

history = model.fit(x_train,
          {'classification': y_train, 'autoencoder': x_train},
          batch_size=batch_size,
          epochs=epochs,
          validation_data= (x_test, {'classification': y_test, 'autoencoder': x_test}),
          verbose=1)
```

```
Epoch 1/1
60000/60000 [==============================] - 41s 677us/step - loss: 0.5086 - classification_loss: 0.4051 - autoencoder_loss: 0.2069 - classification_acc: 0.8755 - autoencoder_batch_counter: 299.7983 - val_loss: 0.2001 - val_classification_loss: 0.1242 - val_autoencoder_loss: 0.1518 - val_classification_acc: 0.9596 - val_autoencoder_batch_counter: 50.1000
```

``autoencoder_batch_counter`` & ``val_autoencoder_batch_counter`` should always be (600, 100) respectively.  These metrics are batch averaged. This does not happen in the Keras implementation.

**Comment by brge17** [comment]
### Problem 2

Batchwise averaging with ``fit_generator`` + Problem 1.

```python
import tensorflow as tf
from tensorflow.python.keras.layers import Input, Dense
from tensorflow.python.keras.models import Model

import numpy as np

class BatchCounter(tf.keras.layers.Layer):

        def __init__(self, name="batch_counter", **kwargs):
            super(BatchCounter, self).__init__(name=name, **kwargs)
            self.stateful = True
            self.batches = tf.keras.backend.variable(value=0, dtype="int32")

        def reset_states(self):
            tf.keras.backend.set_value(self.batches, 0)

        def __call__(self, y_true, y_pred):
            updates = [
                tf.keras.backend.update_add(
                    self.batches, 
                    tf.keras.backend.variable(value=1, dtype="int32"))]
            self.add_update(updates)
            return self.batches

class DummyGenerator(object):
    """ Dummy data generator. """

    def run(self):
        while True:
            yield np.ones((10, 1)), np.zeros((10, 1))

train_gen = DummyGenerator()
val_gen = DummyGenerator()

# Dummy model
inputs = Input(shape=(1,))
outputs = Dense(1)(inputs)
model = Model(inputs=inputs, outputs=outputs)
model.compile(loss="mse", optimizer="adam", metrics=[BatchCounter()])

model.fit_generator(
    train_gen.run(), 
    steps_per_epoch=5, 
    epochs=10, 
    validation_data=val_gen.run(), 
    validation_steps=5)
```

```
5/5 [==============================] - 0s 58ms/step - loss: 2.5346 - batch_counter: 2.2000 - val_loss: 2.5155 - val_batch_counter: 4.0000
Epoch 2/10
5/5 [==============================] - 0s 3ms/step - loss: 2.5029 - batch_counter: 2.2000 - val_loss: 2.4839 - val_batch_counter: 5.0000
Epoch 3/10
5/5 [==============================] - 0s 3ms/step - loss: 2.4713 - batch_counter: 2.6000 - val_loss: 2.4525 - val_batch_counter: 4.0000
Epoch 4/10
5/5 [==============================] - 0s 4ms/step - loss: 2.4400 - batch_counter: 2.4000 - val_loss: 2.4214 - val_batch_counter: 5.0000
Epoch 5/10
5/5 [==============================] - 0s 4ms/step - loss: 2.4090 - batch_counter: 2.2000 - val_loss: 2.3905 - val_batch_counter: 4.0000
Epoch 6/10
5/5 [==============================] - 0s 3ms/step - loss: 2.3782 - batch_counter: 2.4000 - val_loss: 2.3598 - val_batch_counter: 4.0000
Epoch 7/10
5/5 [==============================] - 0s 3ms/step - loss: 2.3476 - batch_counter: 2.8000 - val_loss: 2.3293 - val_batch_counter: 4.0000
Epoch 8/10
5/5 [==============================] - 0s 4ms/step - loss: 2.3173 - batch_counter: 2.4000 - val_loss: 2.2993 - val_batch_counter: 4.0000
Epoch 9/10
5/5 [==============================] - 0s 4ms/step - loss: 2.2873 - batch_counter: 2.8000 - val_loss: 2.2695 - val_batch_counter: 4.0000
Epoch 10/10
5/5 [==============================] - 0s 4ms/step - loss: 2.2577 - batch_counter: 2.0000 - val_loss: 2.2400 - val_batch_counter: 4.0000
```

batch_counter & val_batch_counter should always be 5 in the above logs, (5 steps of 10 samples). They are occasionally 4 or 5. In addition to batch averaging of the metrics. This does not happen in the Keras implementation.

**Comment by brge17** [comment]
### Problem 3
Issues with multi-input/multi-output and batch averaging. This happens for both train and validation metrics.

```python
import tensorflow as tf

from tensorflow.python.keras.datasets import mnist
from tensorflow.python.keras.models import Model
from tensorflow.python.keras.layers import Input, Conv2D, MaxPooling2D, Flatten, Dense, UpSampling2D


class BatchCounter(tf.keras.layers.Layer):

        def __init__(self, name='batch_counter', **kwargs):
            super(BatchCounter, self).__init__(name=name, **kwargs)
            self.stateful = True
            self.batches = tf.keras.backend.variable(value=0, dtype='int32')

        def reset_states(self):
            tf.keras.backend.set_value(self.batches, 0)

        def __call__(self, y_true, y_pred):
            updates = [tf.keras.backend.update_add(self.batches, tf.keras.backend.variable(value=1, dtype='int32'))]
            self.add_update(updates)
            return self.batches


batch_size = 100
num_classes = 10
epochs = 1

# input image dimensions
img_rows, img_cols = 28, 28

# Data
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1).astype('float32') / 255
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1).astype('float32') / 255
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

# Convolutional Encoder
input_img = Input(shape=(img_rows, img_cols, 1))
conv_1 = Conv2D(16, (3, 3), activation='relu', padding='same')(input_img)
pool_1 = MaxPooling2D((2, 2), padding='same')(conv_1)
conv_2 = Conv2D(8, (3, 3), activation='relu', padding='same')(pool_1)
pool_2 = MaxPooling2D((2, 2), padding='same')(conv_2)
conv_3 = Conv2D(8, (3, 3), activation='relu', padding='same')(pool_2)
encoded= MaxPooling2D((2, 2), padding='same')(conv_3)

# Classification
flatten = Flatten()(encoded)
fc = Dense(128, activation='relu')(flatten)
softmax = Dense(num_classes, activation='softmax', name='classification')(fc)

# Decoder
conv_4 = Conv2D(8, (3, 3), activation='relu', padding='same')(encoded)
up_1 = UpSampling2D((2, 2))(conv_4)
conv_5 = Conv2D(8, (3, 3), activation='relu', padding='same')(up_1)
up_2 = UpSampling2D((2, 2))(conv_5)
conv_6 = Conv2D(16, (3, 3), activation='relu')(up_2)
up_3 = UpSampling2D((2, 2))(conv_6)
decoded = Conv2D(1, (3, 3), activation='sigmoid', padding='same', name='autoencoder')(up_3)

model = Model(inputs=input_img, outputs=[softmax, decoded])

model.compile(loss={'classification': 'categorical_crossentropy',
                    'autoencoder': 'binary_crossentropy'},
              loss_weights={'classification': 1.0,
                            'autoencoder': 0.5},
              optimizer='adam',
              metrics={'classification': 'accuracy', 'autoencoder': BatchCounter()})

history = model.fit(x_train,
          {'classification': y_train, 'autoencoder': x_train},
          batch_size=batch_size,
          epochs=epochs,
          validation_data= (x_test, {'classification': y_test, 'autoencoder': x_test}),
          verbose=1)

```



### Source code/logs

**Comment by brge17** [comment]
Related to this issue (was not fixed). https://github.com/tensorflow/tensorflow/issues/19186

**Comment by aselle** [comment]
@fchollet, could you please take a look?

**Comment by briannemsick** [comment]
@pavithrasv could you also take a look at https://github.com/keras-team/keras/pull/10673

The issue is also apparent in tf.keras.

**Comment by pavithrasv** [comment]
Thank you @briannemsick, as @fchollet started reviewing the change I think we should wait for approval from him?

**Comment by brge17** [comment]
@pavithrasv Do you want some help?

I'm very keen on using Stateful Metrics for production.

**Comment by pavithrasv** [comment]
Adding @raymond-yuan who is looking into this issue.

**Comment by raymond-yuan** [comment]
The batch averaging issue has been resolved, looking into the occasional metric not updating issue

**Comment by brge17** [comment]
Thanks for the response @raymond-yuan. Assuming it is still an internal google branch.

Should I make a new issue for: https://github.com/tensorflow/tensorflow/pull/21071

This is the ``TensorBoard`` issue with Stateful Metrics. It was merged into Keras https://github.com/keras-team/keras/pull/10673

**Comment by brge17** [comment]
@raymond-yuan I think the progress bar is still busted for ``fit_generator``:

https://github.com/tensorflow/tensorflow/pull/20650/files

Looks like ``keras`` and ``tf.keras`` are out sync.

**Comment by raymond-yuan** [comment]
Hey @brge17 sorry for the late response

> Should I make a new issue for: #21071
Yes please make a new issue for #21071

If there's nothing directly related to this issue, I'll go ahead and close it?

**Comment by brge17** [comment]
Is the metric not updating issue fixed?

**Comment by raymond-yuan** [comment]
@pavithrasv will be looking into that issue!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by brge17** [comment]
@pavithrasv any update on the missing/overcounting batches issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @pavithrasv: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by deeb02** [comment]
Closing as stale.
